Firestore/core/src/firebase/firestore/local/leveldb_migrations.cc

```diff
@@ -110,6 +113,71 @@ void ClearQueryCache(leveldb::DB* db) {
   transaction.Commit();
 }
 
+void RemoveMutationDocuments(LevelDbTransaction *transaction,
+                             absl::string_view user_id,
+                             int32_t last_acknowledged_batch_id) {
+  LevelDbDocumentMutationKey doc_key;
+  std::string prefix = LevelDbDocumentMutationKey::KeyPrefix(user_id);
+
+  auto it = transaction->NewIterator();
+  it->Seek(prefix);
+  for (; it->Valid() && absl::StartsWith(it->key(), prefix); it->Next()) {
+    HARD_ASSERT(doc_key.Decode(it->key()),
+                "Failed to decode document mutation key");
+    if (doc_key.batch_id() <= last_acknowledged_batch_id) {
+      transaction->Delete(it->key());
+    }
+  }
+}
+
+void RemoveMutationBatches(LevelDbTransaction *transaction,
+                           absl::string_view user_id,
+                           int32_t last_acknowledged_batch_id) {
+  LevelDbMutationKey key;
+  std::string mutations_key = LevelDbMutationKey::KeyPrefix(user_id);
+  std::string last_key =
+      LevelDbMutationKey::Key(user_id, last_acknowledged_batch_id);
+  auto it = transaction->NewIterator();
+  it->Seek(mutations_key);
+  for (; it->Valid() && absl::StartsWith(it->key(), mutations_key) &&
```

Comment:
Doesn't `it->key() <= last_key` imply `absl::StartsWith(it->key(), mutations_key)`?

Reply:
Yeah.